test/functional/legacy/080_substitute_spec.lua

```diff
@@ -0,0 +1,177 @@
+    eq('e\\\re', eval([[substitute('eEe', 'E', ]] ..
+      [['\=substitute(submatch(0), ".", "\\\\]]..'\r'..[[", "")', '')]]))
+    eq('f\rf', eval([[substitute('fFf', 'F', ]] ..
+      [['\=substitute(submatch(0), ".", "\\r", "")', '')]]))
+    eq('j\nj', eval([[substitute('jJj', 'J', ]] ..
+      [['\=substitute(submatch(0), ".", "\\n", "")', '')]]))
+    eq('k\rk', eval([[substitute('kKk', 'K', ]] ..
+      [['\=substitute(submatch(0), ".", "\r", "")', '')]]))
+    eq('l\nl', eval([[substitute('lLl', 'L', ]] ..
+      [['\=substitute(submatch(0), ".", "\n", "")', '')]]))
+  end)
+
+  it('with submatch() (TEST_5)', function()
+    execute('set magic&')
+    execute('set cpo&')
+    eq('A123456789987654321', eval([[substitute('A123456789', ]] ..
+      [['A\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)', ]] ..
+      [['\=submatch(0) . submatch(9) . submatch(8) . submatch(7) . ]] ..
+      [[submatch(6) . submatch(5) . submatch(4) . submatch(3) . ]] ..
+      [[submatch(2) . submatch(1)', '')]]))
+    eq("[['A123456789'], ['9'], ['8'], ['7'], ['6'], ['5'], ['4'], ['3'], " ..
+      "['2'], ['1']]", eval([[substitute('A123456789', ]] ..
+      [['A\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)', ]] ..
+      [['\=string([submatch(0, 1), submatch(9, 1), submatch(8, 1), ]] ..
+      [[submatch(7, 1), submatch(6, 1), submatch(5, 1), submatch(4, 1), ]] ..
+      [[submatch(3, 1), submatch(2, 1), submatch(1, 1)])', '')]]))
+  end)
+
+  it('with "set cpo±=/" (TEST_6)', function()
+    execute('set magic&')
+    execute('set cpo+=/')
+    eq('a', eval("substitute('A', 'A', 'a', '')"))
+    eq('%', eval("substitute('B', 'B', '%', '')"))
+    execute('set cpo-=/')
+    eq('c', eval("substitute('C', 'C', 'c', '')"))
+    eq('%', eval("substitute('D', 'D', '%', '')"))
+  end)
+
+  it('with submatch or \\ze (TEST_7)', function()
+    execute('set magic&')
+    execute('set cpo&')
+    eq('A\rA', eval("substitute('A\rA', 'A.', '\\=submatch(0)', '')"))
+    eq('B\nB', eval([[substitute("B\nB", 'B.', '\=submatch(0)', '')]]))
+    eq("['B\n']B",
+      eval([[substitute("B\nB", 'B.', '\=string(submatch(0, 1))', '')]]))
+    eq('-abab', eval([[substitute('-bb', '\zeb', 'a', 'g')]]))
+    eq('c-cbcbc', eval([[substitute('-bb', '\ze', 'c', 'g')]]))
+  end)
+
+  it('with \\zs and \\ze (TEST_10)', function()
+    execute('set magic&')
+    execute('set cpo&')
+    eq('a1a2a3a', eval([[substitute('123', '\zs', 'a', 'g')]]))
+    eq('aaa', eval([[substitute('123', '\zs.', 'a', 'g')]]))
+    eq('1a2a3a', eval([[substitute('123', '.\zs', 'a', 'g')]]))
+    eq('a1a2a3a', eval([[substitute('123', '\ze', 'a', 'g')]]))
+    eq('a1a2a3', eval([[substitute('123', '\ze.', 'a', 'g')]]))
+    eq('aaa', eval([[substitute('123', '.\ze', 'a', 'g')]]))
+    eq('aa2a3a', eval([[substitute('123', '1\|\ze', 'a', 'g')]]))
+    eq('1aaa', eval([[substitute('123', '1\zs\|[23]', 'a', 'g')]]))
+  end)
+end)
+
+describe(':substitue', function()
+  before_each(clear)
```

Comment:
There might be tests that require *more than this*. It is described in some comment on the issues that describe writing and conversion of tests here on GH. I don't know where exactly.